core/src/main/java/com/linecorp/armeria/client/Bootstraps.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client;
+
+import static com.linecorp.armeria.common.SessionProtocol.httpAndHttpsValues;
+
+import java.lang.reflect.Array;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.util.Set;
+
+import com.linecorp.armeria.common.SerializationFormat;
+import com.linecorp.armeria.common.SessionProtocol;
+import com.linecorp.armeria.common.annotation.Nullable;
+
+import io.netty.bootstrap.Bootstrap;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelInitializer;
+import io.netty.channel.ChannelOption;
+import io.netty.channel.EventLoop;
+import io.netty.channel.unix.DomainSocketAddress;
+import io.netty.handler.ssl.SslContext;
+
+final class Bootstraps {
+    private final Bootstrap[][] inetBootstraps;
+    @Nullable
+    private final Bootstrap[][] unixBootstraps;
+
+    private final int connectTimeoutMillis;
+
+    private final EventLoop eventLoop;
+
+    private final SslContext sslCtxHttp1Only;
+
+    private final SslContext sslCtxHttp1Or2;
+
+    Bootstraps(HttpClientFactory clientFactory, EventLoop eventLoop, SslContext sslCtxHttp1Only,
+               SslContext sslCtxHttp1Or2) {
+        this.eventLoop = eventLoop;
+        this.sslCtxHttp1Only = sslCtxHttp1Only;
+        this.sslCtxHttp1Or2 = sslCtxHttp1Or2;
+        final Bootstrap inetBaseBootstrap = clientFactory.newInetBootstrap();
+        final Bootstrap unixBaseBootstrap = clientFactory.newUnixBootstrap();
+        inetBootstraps = newBootstrapMap(inetBaseBootstrap, clientFactory, eventLoop);
+        if (unixBaseBootstrap != null) {
+            unixBootstraps = newBootstrapMap(unixBaseBootstrap, clientFactory, eventLoop);
+        } else {
+            unixBootstraps = null;
+        }
+        connectTimeoutMillis = (Integer) inetBaseBootstrap.config().options()
+                                                          .get(ChannelOption.CONNECT_TIMEOUT_MILLIS);
+    }
+
+    Bootstrap get(SocketAddress remoteAddress, SessionProtocol desiredProtocol,
+                  SerializationFormat serializationFormat) {
+        if (!httpAndHttpsValues().contains(desiredProtocol)) {
+            throw new NullPointerException("Unsupported session protocol: " + desiredProtocol);
```

Comment:
`IllegalArgumentException`?

Reply:
Oh good, I forgot to change the exception I set NPE to something else. 😂 Thanks